solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs

```diff
@@ -295,26 +295,123 @@ impl TokenTransferValidationContext for IbcStorage<'_, '_> {
 
     fn send_coins_validate(
         &self,
-        _from_account: &Self::AccountId,
-        _to_account: &Self::AccountId,
-        _coin: &PrefixedCoin,
+        from_account: &Self::AccountId,
+        to_account: &Self::AccountId,
+        coin: &PrefixedCoin,
     ) -> Result<(), TokenTransferError> {
+        /*
+           Should have the following accounts
+           - token program
+           - token account
+           - escrow account ( with seeds as portId, channelId and denom )
+           - token mint
+
+           If sending tokens from escrow then,
+           - mint authority should be present
+           - from account should match escrow
+           - to account should match token account
+
+          If sending tokens to escrow then,
+           - sender should be present
+           - sender should be signer
+           - to account should match escrow
+           - from account should match token account
+        */
+        let store = self.borrow();
+        let accounts = &store.accounts;
+        if accounts.token_program.is_none() ||
+            accounts.token_mint.is_none() ||
+            accounts.escrow_account.is_none() ||
+            accounts.token_account.is_none()
+        {
+            return Err(TokenTransferError::ParseAccountFailure);
+        }
+        let escrow_account = &accounts.escrow_account.clone().unwrap(); // we check before if it exists so the unwrap is infallible
+        let token_account = &accounts.token_account.clone().unwrap(); // we check before if it exists so the unwrap is infallible
+        if matches!(from_account, AccountId::Escrow(_)) {
+            let from = from_account
+                .get_escrow_account(coin.denom.base_denom.as_str())
+                .map_err(|_| TokenTransferError::ParseAccountFailure)?;
+            let to: Pubkey = to_account.try_into().unwrap(); // Infallible
+            if accounts.mint_authority.is_none() {
+                return Err(TokenTransferError::ParseAccountFailure);
+            }
+            if !from.eq(escrow_account.key) || !to.eq(token_account.key) {
+                return Err(TokenTransferError::ParseAccountFailure);
+            }
+        } else {
+            let from: Pubkey = from_account.try_into().unwrap(); // Infallible
+            let to = from_account
+                .get_escrow_account(coin.denom.base_denom.as_str())
+                .map_err(|_| TokenTransferError::ParseAccountFailure)?;
+            if accounts.sender.is_none() {
+                return Err(TokenTransferError::ParseAccountFailure);
+            }
+            if !accounts.sender.clone().unwrap().is_signer {
+                return Err(TokenTransferError::ParseAccountFailure);
+            }
+            if !from.eq(escrow_account.key) || !to.eq(token_account.key) {
+                return Err(TokenTransferError::ParseAccountFailure);
+            }
+        }
         Ok(())
```

Comment:
How about `if let` to avoid the unwraps?  Strictly speaking they
aren’t infallible.  Also it might be more concise to have an `ok`
flag.  Lastly, whenever you feel you need to clone an Option, consider
using Option::as_ref instead.

```suggestion
        let ok = if let AccountId::Signer(ref to) = to_account {
            let from = from_account
                .get_escrow_account(coin.denom.base_denom.as_str())
                .map_err(|_| TokenTransferError::ParseAccountFailure)?;
            accounts.mint_authority.is_some() &&
                from.eq(escrow_account.key) &&
                to.eq(token_account.key)
        } else if let AccountId::Signer(ref from) = from_account {
            let to = from_account
                .get_escrow_account(coin.denom.base_denom.as_str())
                .map_err(|_| TokenTransferError::ParseAccountFailure)?;
            accounts.sender.is_some() &&
                accounts.sender.as_ref().map_or(false, |acc| acc.is_signer) &&
                from.eq(escrow_account.key) &&
                to.eq(token_account.key)
        } else {
            false
        };
        if ok { Ok(()) } else { Err(TokenTransferError::ParseAccountFailure) }
```